core/src/main/java/jenkins/security/RSAConfidentialKey.java

```diff
@@ -0,0 +1,88 @@
+package jenkins.security;
+
+import org.apache.commons.codec.binary.Base64;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.KeyFactory;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.PrivateKey;
+import java.security.SecureRandom;
+import java.security.interfaces.RSAPrivateCrtKey;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.RSAPublicKeySpec;
+
+/**
+ * RSA public/private key pair as {@link ConfidentialKey}.
+ *
+ * <p>
+ * As per the design principle of {@link ConfidentialKey}, not exposing {@link PrivateKey} directly.
+ * Define subtypes for different use cases.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+public abstract class RSAConfidentialKey extends ConfidentialKey {
+    private volatile RSAPrivateKey priv;
```

Comment:
:-1: on multiple volatiles with a lock used to guard write access. This pattern is the root of much evil with threading in Jenkins